destination/writer/writer.go

```diff
@@ -0,0 +1,163 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package writer
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	sdk "github.com/conduitio/conduit-connector-sdk"
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/mongo"
+)
+
+const (
+	// idFieldName contains default reserved primary key from MongoDB.
+	idFieldName = "_id"
+
+	// setCommand contains command, that used during Update query.
+	setCommand = "$set"
+)
+
+// Writer implements a writer logic for Mongo destination.
+type Writer struct {
+	collection *mongo.Collection
+}
+
+// NewWriter creates new instance of the Writer.
+func NewWriter(ctx context.Context, collection *mongo.Collection) *Writer {
+	writer := &Writer{
+		collection: collection,
+	}
+
+	return writer
+}
+
+// Write inserts a sdk.Record into a Destination.
+func (w *Writer) Write(ctx context.Context, record sdk.Record) error {
+	if err := sdk.Util.Destination.Route(ctx, record,
+		w.insert,
+		w.update,
+		w.delete,
+		w.insert,
+	); err != nil {
+		return fmt.Errorf("route %s: %w", record.Operation, err)
+	}
+
+	return nil
+}
+
+// Close closes the underlying db connection.
+func (w *Writer) Close(ctx context.Context) error {
+	err := w.collection.Database().Client().Disconnect(ctx)
+	if err != nil {
+		return fmt.Errorf("close collection: %w", err)
+	}
+
+	return nil
+}
+
+func (w *Writer) insert(ctx context.Context, record sdk.Record) error {
+	payload := make(sdk.StructuredData)
+	if err := json.Unmarshal(record.Payload.After.Bytes(), &payload); err != nil {
+		return fmt.Errorf("unmarshal payload: %w", err)
+	}
+
+	if _, err := w.collection.InsertOne(ctx, payload); err != nil {
+		return fmt.Errorf("insert data: %w", err)
+	}
+
+	return nil
+}
+
+func (w *Writer) update(ctx context.Context, record sdk.Record) error {
+	payload, err := parsePayload(record)
+	if err != nil {
+		return fmt.Errorf("parse payload: %w", err)
+	}
+
+	ids, err := parseKeys(record)
+	if err != nil {
+		return fmt.Errorf("parse keys: %w", err)
+	}
+
+	filter := generateBsonFromMap(ids)
+	body := generateBsonFromMap(payload)
+
+	if _, err := w.collection.UpdateOne(ctx,
+		filter,
+		bson.D{{
+			Key:   setCommand,
+			Value: body,
+		}},
+	); err != nil {
+		return fmt.Errorf("update payload into destination: %w", err)
+	}
+
+	return nil
+}
+
+func (w *Writer) delete(ctx context.Context, record sdk.Record) error {
+	ids, err := parseKeys(record)
+	if err != nil {
+		return fmt.Errorf("parse keys: %w", err)
+	}
+
+	filter := generateBsonFromMap(ids)
+
+	if _, err := w.collection.DeleteOne(ctx,
+		filter,
+	); err != nil {
+		return fmt.Errorf("delete data from destination: %w", err)
+	}
+
+	return nil
+}
+
+// parsePayload is parsing Payload.After from record and deleting key from payload (we using it from Key field).
+func parsePayload(data sdk.Record) (map[string]any, error) {
```

Comment:
Let's get rid of this function, and call the `json.Unmarshal` directly. And rename `parsed` to `payload`.

Reply:
fixed